### PR TITLE
RD-1758 / RD-1774 & RD-1762 - Add support for type: 'attribute' and operators: 'contains', 'not_contains', 'starts_with', 'ends_with'

### PR DESCRIPTION
**Pull request by qooban**
### Changes
* Added `MultipleStringValuesInput` component allowing to type in multiple string values
* Extracted `isAnyOperator` common function
* Used `MultipleStringValuesInput` component in `AttributeValueInput` to allow full support for type: `attribute` and operators: `contains`, `not_contains`, `starts_with`, `ends_with`
* Extracted `getTranslation` and `getPlaceholderTranslation` common functions

Tests will come from: https://github.com/cloudify-cosmo/cloudify-stage/pull/1313

### Video
https://user-images.githubusercontent.com/5202105/114687646-42c44d80-9d14-11eb-839f-b22029a912c1.mp4

### CODE REVIEW

**Comment by Gelio** [comment]
I have meetings till the end of the day today, so I'll read through the PR tomorrow :smile:

File: app/translations/en.json [hunk 1]
```diff
@@ -531,7 +531,8 @@
                     "inputsPlaceholders": {
                         "blueprintId": "Select blueprints",
                         "siteName": "Select site names",
-                        "createdBy": "Select usernames"
+                        "createdBy": "Select usernames",
+                        "multipleStrings": "Type in texts"
```
Comment (kubama) [inline app/translations/en.json:535]: How about "Type in values" or "Enter values"?

File: widgets/common/src/filters/inputs/common.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import type { FilterRuleOperator } from '../types';
+import { FilterRuleOperators } from '../types';
+
+// eslint-disable-next-line import/prefer-default-export
+export function isAnyOfOrNotAnyOfOperator(operator: FilterRuleOperator) {
```
Comment (kubama) [inline widgets/common/src/filters/inputs/common.ts:5]: Quite hard to read and even a little bit hard to understand function name. How about shortening to `isAnyOperator`? Or, given the context in which it is used `isExactValueOperator`?

File: widgets/common/src/filters/inputs/AttributeValueInput.tsx [hunk 3]
```diff
@@ -1,9 +1,10 @@
-import { castArray } from 'lodash';
 import type { FunctionComponent } from 'react';
 
 import DynamicDropdown from '../../DynamicDropdown';
-import { LabelsFilterRuleOperators } from '../types';
 import type { CommonAttributeValueInputProps } from './types';
+import { isAnyOfOrNotAnyOfOperator } from './common';
+import { i18n, i18nPlaceholdersPrefix } from '../consts';
```
Comment (kubama) [inline widgets/common/src/filters/inputs/AttributeValueInput.tsx:6]: Looks like importing `i18n` and `i18nPlaceholdersPrefix` or `i18nPrefix` is a common pattern. Instead I'd consider exporting a function (or actually two functions) that would add a prefix and delegate to `i18n.t`. It could be placed in `common.ts` and `consts.ts` would be not needed.

File: app/translations/en.json (hunk 1 above)
Comment (qooban) [reply app/translations/en.json]: I was debating with myself what to put here. I like the first one, so changed it to "Type in values".

File: widgets/common/src/filters/inputs/common.ts (hunk 2 above)
Comment (qooban) [reply widgets/common/src/filters/inputs/common.ts]: I was sure that somebody will point that out 🙂 
I like the first suggestion, changed.

File: widgets/common/src/filters/inputs/AttributeValueInput.tsx (hunk 3 above)
Comment (qooban) [reply widgets/common/src/filters/inputs/AttributeValueInput.tsx]: Good idea. That involved a bit bigger change, so many files updated.

NOTE: There are 2 `common.ts` files right now. I'm OK with having two as they do their job very precisely.

**Review by kubama — APPROVED** [review]

**Comment by qooban** [comment]
Started system test execution: https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/456/

Waiting for review from @Gelio and I hope to merge this PR today.

**Comment by Gelio** [comment]
I'm beginning with the review :eyes: I hope to be faster than Jenkins :racehorse:

File: widgets/common/src/filters/RuleRowTypeDropdown.tsx [hunk 4]
```diff
@@ -13,7 +13,7 @@ interface RuleRowTypeDropdownProps {
 const ruleRowTypeOptions = (() => {
     return Object.values(FilterRuleRowType).map(
         (ruleType): DropdownItemProps => ({
-            text: i18n.t(`${i18nPrefix}.ruleTypesLabels.${camelCase(ruleType)}`),
+            text: getTranslation(`ruleTypesLabels.${camelCase(ruleType)}`),
```
Comment (Gelio) [inline widgets/common/src/filters/RuleRowTypeDropdown.tsx:16]: I think I like this `getTranslation` pattern more than exporting the prefix :smile: Looks shorter :slightly_smiling_face: 

Good suggestion @kubama :+1:

File: widgets/common/src/filters/inputs/common.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import type { FilterRuleOperator } from '../types';
+import { FilterRuleOperators } from '../types';
+import { getTranslation } from '../common';
+
+export function isAnyOperator(operator: FilterRuleOperator) {
+    return ([FilterRuleOperators.AnyOf, FilterRuleOperators.NotAnyOf] as FilterRuleOperator[]).includes(operator);
```
Comment (Gelio) [inline widgets/common/src/filters/inputs/common.ts:6]: Optional:

Using `as` is less safe than a type annotation for a variable. For example, the following works:

```diff
-    return ([FilterRuleOperators.AnyOf, FilterRuleOperators.NotAnyOf] as FilterRuleOperator[]).includes(operator);
+    return ([FilterRuleOperators.AnyOf, FilterRuleOperators.NotAnyOf, 'abc'] as FilterRuleOperator[]).includes(operator);
```

Whereas, the following reports an error:

```ts
    const anyOperators: FilterRuleOperator[] = [FilterRuleOperators.AnyOf, FilterRuleOperators.NotAnyOf, 'abc'];
    return anyOperators.includes(operator);
```

image

Thus, I suggest refactoring this to the following:

```suggestion
    const anyOperators: FilterRuleOperator[] = [FilterRuleOperators.AnyOf, FilterRuleOperators.NotAnyOf];
    return anyOperators.includes(operator);
```

This is optional here, but have it in mind in the future :smile: When it's not too much hassle to create a variable, do that to be more type-safe :slightly_smiling_face: 

A bit more info: https://stackoverflow.com/a/54734992/4874344

In short, `as` is potentially unsafe, as it will accept some type-mismatches :slightly_smiling_face: It won't accept anything, e.g. `[] as string` would not work, since the types are very different, but it's worth keeping in mind that `as` has more "power"

File: widgets/common/src/filters/inputs/common.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import type { FilterRuleOperator } from '../types';
+import { FilterRuleOperators } from '../types';
+import { getTranslation } from '../common';
+
+export function isAnyOperator(operator: FilterRuleOperator) {
```
Comment (Gelio) [inline widgets/common/src/filters/inputs/common.ts:5]: Optional:

I know it's been discussed already, but I still find the name not matching the behavior :smile: _is any operator_ to me sounds like a function that would check if some string is an operator :smile: 

I don't have a great name here, but perhaps we could not name it by what it does mechanically, but rather by what's the intent when calling this function. I see when it's called, we display `DynamicDropdown`. Thus, how about `shouldAutocompleteKnownValues`? :smile:

File: widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+const MultipleStringValuesInput: FunctionComponent<MultipleStringValuesInputProps> = ({
+    name,
+    onChange,
+    placeholder,
+    value
+}) => {
+    const { Dropdown } = Stage.Basic;
+    const options = useMemo(() => value.map(element => ({ text: element, value: element })), [value]);
```
Comment (Gelio) [inline widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx:17]: Optional:

Is there a type we could use for the return type of the mapping function to get a better error message if it does not match the expected type? I think we could do

```suggestion
    const options = useMemo(() => value.map((element): DropdownItemProps => ({ text: element, value: element })), [value]);
```

File: widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx [hunk 8]
```diff
@@ -0,0 +1,35 @@
+import { FunctionComponent, useMemo } from 'react';
+
+interface MultipleStringValuesInputProps {
+    name: string;
+    onChange: (newValue: string[]) => void;
+    placeholder: string;
+    value: string[];
```
Comment (Gelio) [inline widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx:7]: Optional:

I prefer using plurals for arrays :smile: 

```suggestion
    values: string[];
```

**Review by Gelio — APPROVED** [review]
Left minor comments, did not run it in the UI, but overall, looks good :smile: I like the translation-function pattern :+1: Seems better than exposing the prefixes :slightly_smiling_face:

File: widgets/common/src/filters/inputs/common.ts (hunk 5 above)
Comment (qooban) [reply widgets/common/src/filters/inputs/common.ts]: Thanks for the description, appreciate that. Applied your suggestion.

File: widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx (hunk 7 above)
Comment (qooban) [reply widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx]: Sure, make sense. Applied.

File: widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx (hunk 8 above)
Comment (qooban) [reply widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx:8]: Hmm... If we would do it so here, then I think we should do it as well in the other places:
image

As this will be a bit bigger change, I would leave it for now and maybe consider changing it later.

File: widgets/common/src/filters/inputs/common.ts (hunk 6 above)
Comment (qooban) [reply widgets/common/src/filters/inputs/common.ts:5]: Hmm... I'm not sure about the suggested name..

Another option would to change it to `isCommonRuleOperator` as `AnyOf` and `NotAnyOf` are defined inside `CommonRuleOperator` type.

Then we could probably even more change the function to have something like:
```
export function isCommonRuleOperator(operator: FilterRuleOperator) {
    return Object.keys(CommonRuleOperator).includes(operator);
}
```
(for that we would need to `export enum CommonRuleOperator` from `filters/types.ts`)

Anyways, for this PR I would leave it as it is. If we all agree to any specific change, I'd like to apply it in the later phase.

Comment (Gelio) [reply widgets/common/src/filters/inputs/common.ts:5]: I see how we could leverage `CommonRuleOperator`, but I'm not sure if it will always stay like that - if we add another `CommonRuleOperator`, will it automatically mean that we use the `DynamicDropdown` for it? Probably not :smile: Thus, I would be cautious going in that direction

I don't have a better suggestion than I gave in the initial comment, so we can ignore the matter for now :slightly_smiling_face:

File: widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx (hunk 8 above)
Comment (Gelio) [reply widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx:8]: I see :smile: We can ignore the matter here, but generally, does it matter that we always make changes in a single step? I don't see it as problematic that we would have it a prop named `values` and we would pass `data.value` to it :slightly_smiling_face: I find it easier to make small improvements when it's possible, even if that means that we break some existing pattern (that we can agree can be improved)

File: widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx [hunk 9]
```diff
@@ -1,4 +1,5 @@
 import { FunctionComponent, useMemo } from 'react';
+import { DropdownItemProps } from 'semantic-ui-react';
```
Comment (Gelio) [inline widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx:2]: Optional: `import type` could help, but let's save time, we can do without it here and merge the PR with this small blunder :smile:

**Review by Gelio — APPROVED** [review]

File: widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx (hunk 9 above)
Comment (qooban) [reply widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx:2]: Oops, yes, we need RD-2000 🙂

Changed.

**Comment by qooban** [comment]
Previous test execution failed on Filters widget test. I applied the fix within https://github.com/cloudify-cosmo/cloudify-stage/pull/1312/commits/732b70da9335cba256231f667fc1f94eccb21361 . More test changes are already started and will be applied later this or next week within: https://github.com/cloudify-cosmo/cloudify-stage/pull/1313

I don't want to wait for another test execution, so I tested it locally:
image

File: widgets/common/src/filters/inputs/common.ts (hunk 6 above)
Comment (qooban) [reply widgets/common/src/filters/inputs/common.ts:5]: OK, so let's ignore it.

**Review by Gelio — DISMISSED** [review]
Good thing we've detected the `filters_spec` issue before merging :smile:

File: widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx (hunk 8 above)
Comment (qooban) [reply widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx:8]: I'm not sure if I understand the message entirely. In general more often I'm leaning towards having consistent application, instead of perfectly optimized parts. Maybe that is something to discuss in the weekly 🙂 Maybe that's just the preferences difference which we can live with 🙂

File: test/cypress/integration/widgets/filters_spec.ts [hunk 10]
```diff
@@ -17,6 +17,12 @@ describe('Filters widget', () => {
         cy.get('.loading').should('not.exist');
     });
 
+    function typeAttributeRuleValue(value: string) {
+        cy.get('[name=ruleValue]')
+            .click()
+            .within(() => cy.get('input').type(`${value}{enter}`).blur());
```
Comment (kubama) [inline test/cypress/integration/widgets/filters_spec.ts:23]:
```suggestion
            .find('input').type(`${value}{enter}`).blur();
```

Comment (qooban) [reply test/cypress/integration/widgets/filters_spec.ts]: I know that's not the the first time you remind me about it 🙂 
Changed. `filters_spec` tests passed.

**Comment by qooban** [comment]
Did one small change related to https://github.com/cloudify-cosmo/cloudify-stage/pull/1312#discussion_r613938545 .
`filters_spec` tests passed locally, again, so I'm waiting for final approvals and I'm going to merge it then.

File: widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx (hunk 8 above)
Comment (Gelio) [reply widgets/common/src/filters/inputs/MultipleStringValuesInput.tsx:8]: Sure, I've added a point about it :slightly_smiling_face:

**Review by Gelio — APPROVED** [review]

**Review by kubama — APPROVED** [review]
